### PR TITLE
feat(TDP-2807): search bar

### PR DESCRIPTION
**Pull request by frassinier**
**Link to the JIRA issue**
https://jira.talendforge.org/browse/TDP-2807

**Please check if the PR fulfills these requirements**
- [x] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] The code coverage on new code is > 75 % for backend and > 95% for frontend
- [x] The new code does not introduce new technical issues (sonar / eslint)
- [ ] Functional tests have been performed

**Please check the browsers you've tested on**
- [ ] Chrome, Firefox, Safari, Edge, IE11
- [ ] No, that's bad, this PR should not be merged !
- [ ] No, and no need to (backend changes only)

### CODE REVIEW

**Comment by jsomsanith** [comment]
Oops, I approved but didn't want to, don't know how to cancel it

File: dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js [hunk 1]
```diff
@@ -47,24 +79,119 @@ export default class AppHeaderBarCtrl {
...
+						};
+					}),
+				};
+			});
 	}
 
 	adaptContent() {
+		const search = this.appSettings.views.appheaderbar.search ?
+			this.adaptSearch() :
+			{};
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: This should be `null`/`undefined` if we don't have a search config. Otherwise we have the search input that is rendered.

File: dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js [hunk 2]
```diff
@@ -15,14 +15,21 @@ const NAV_ITEM = 'navItem';
 const DROPDOWN = 'dropdown';
 
 export default class AppHeaderBarCtrl {
-	constructor($element, $translate, appSettings, SettingsActionsService) {
+	constructor($element, $translate, appSettings, state, SettingsActionsService, SearchService) {
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: Remove `state`and `SearchService`

File: dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js [hunk 3]
```diff
@@ -47,24 +79,119 @@ export default class AppHeaderBarCtrl {
...
+
+		// onToggle
+		const onToggleAction = this.appSettings.actions[searchSettings.onToggle];
+		this.searchOnToggle = this.settingsActionsService.createDispatcher(onToggleAction);
+
+		// onBlur
+		const onBlurAction = this.appSettings.actions[searchSettings.onBlur];
+		const onBlurActionDispatcher = this.settingsActionsService.createDispatcher(onBlurAction);
+		this.searchOnBlur = (event) => {
+			if (!this.state.search.searchInput) {
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: `if (!this.searchInput)`

File: dataprep-webapp/src/app/components/widgets-containers/widgets-containers-module.js [hunk 4]
```diff
@@ -35,7 +37,9 @@ angular.module(MODULE_NAME,
 		'react',
 		'pascalprecht.translate',
 		FOLDER_MODULE,
+		SEARCH_MODULE,
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/components/widgets-containers/widgets-containers-module.js]: Remove this dependency (not used here anymore)
Can you remove the `FOLDER_MODULE` dependency too ? This is not used neither

File: dataprep-webapp/src/app/components/widgets-containers/layout/layout-container.js [hunk 5]
```diff
@@ -15,7 +15,10 @@ const LayoutContainer = {
 	template: `
 		<div class="app">
 			<div class="header">
-				<app-header-bar><app-header-bar/>
+				<app-header-bar search-toggle="$ctrl.state.search.searchToggle"
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/components/widgets-containers/layout/layout-container.js]: Since this has a state, I think we should refactor the layout to include multi transclusion, so the home component would be responsible to instanciate the header and the side-panel, and transclude them in layout.
But this is just an idea, let's make another PR for that

File: dataprep-webapp/src/assets/config/app-settings.json [hunk 6]
```diff
@@ -395,6 +410,25 @@
         "method": "removeFolder",
         "args": []
       }
+    },
+    "search:toggle": {
```
Comment (amaalej) [inline dataprep-webapp/src/assets/config/app-settings.json]: how about calling it ```appheaderbar:search:toggle``` that would be more precise, no?

File: dataprep-webapp/src/assets/config/app-settings.json [hunk 7]
```diff
@@ -395,6 +410,25 @@
...
+      "id": "search:toggle",
+      "name": "Toggle search input",
+      "icon": "talend-search",
+      "type": "@@search/TOGGLE"
+    },
+    "search:all": {
```
Comment (amaalej) [inline dataprep-webapp/src/assets/config/app-settings.json]: the same here.

File: dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js [hunk 8]
```diff
@@ -32,10 +39,35 @@ export default class AppHeaderBarCtrl {
...
+					searchConfiguration.items = [];
+				}
+				else {
+					delete searchConfiguration.onToggle;
+				}
+			}
+			else if (changes.searchInput) {
+				const searchInput = changes.searchInput.currentValue;
+				searchConfiguration.value = searchInput;
+				if (!searchInput.length) {
```
Comment (amaalej) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: ``` if(!searchInput) ``` is enough no?

File: dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js [hunk 9]
```diff
@@ -47,24 +79,119 @@ export default class AppHeaderBarCtrl {
...
+			onChange: this.searchOnChange,
+			onSelect: this.searchOnToggle && this.searchOnSelect,
+		};
+	}
+
+	_adaptSearchResults(searchResults) {
+		return this.searchAvailableInventoryTypes
+			.filter((inventoryType) => {
+				return searchResults.some((result) => {
+					return result.inventoryType === inventoryType.title;
```
Comment (amaalej) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: Arrow functions no?

File: dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js (hunk 3 above)
Comment (frassinier) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: :white_check_mark:

File: dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js [hunk 10]
```diff
@@ -47,24 +79,119 @@ export default class AppHeaderBarCtrl {
...
+			onClick: this.settingsActionsService.createDispatcher(clickAction),
+		};
+	}
+
+	adaptSearch() {
+		const searchSettings = this.appSettings.views.appheaderbar.search;
+
+		// onToggle
+		const onToggleAction = this.appSettings.actions[searchSettings.onToggle];
+		this.searchOnToggle = this.settingsActionsService.createDispatcher(onToggleAction);
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: :white_check_mark:

File: dataprep-webapp/src/app/components/widgets-containers/widgets-containers-module.js (hunk 4 above)
Comment (frassinier) [inline dataprep-webapp/src/app/components/widgets-containers/widgets-containers-module.js]: :white_check_mark:

File: dataprep-webapp/src/app/services/search/search-module.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * @ngdoc object
+ * @name data-prep.services.search
+ * @description This module contains the services to search
+ */
+angular.module(MODULE_NAME,
+	[
+		SERVICES_DOCUMENTATION_MODULE,
+		SERVICES_EASTER_EGGS_MODULE,
+		SERVICES_INVENTORY_MODULE,
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/services/search/search-module.js]: Can you 
* rename InventoryService to SearchInventoryService
* rename DocumentationService to SearchDocumentationService
* move the 2 folders into services/search

File: dataprep-webapp/src/app/services/search/search-service.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+			results = results.concat(response);
+		});
+
+		// if results (doc + inventory) are empty, we create an empty array
+		// the no-result message is based on the definition of results.
+		// It must be an empty array to show the message.
+		return $q
+			.all([inventoryPromise, documentationPromise])
+			.then(() => {
+				return (state.search.searchInput === searchInput) ? results : [];
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/services/search/search-service.js]: I don't think this is the right place for this check. Because it is based on a state, but this service can possibly by called in any places with any value that is not in `state.search.searchInput`. This is typically something to be managed by the entity that set the value into `state.search.searchInput.`
* just return the result
* remove `state` dependency
* in `search-actions-service.js`: 
```
.then(results => {
	if (state.search.searchInput === searchInput) {
		this.stateService.setSearchResults(results));
	}
}
```

File: dataprep-webapp/src/app/components/search/documentation/documentation-search-controller.spec.js [hunk 13]
```diff
@@ -11,55 +11,57 @@
...
-            component.currentInput = 'other';
-            scope.$digest();
-
-            //then
-            expect(component.results).not.toBe(results);
-        }));
-    });
+describe('Documentation Search controller', () => {
+	let scope;
+	let component;
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/search/documentation/documentation-search-controller.spec.js]: :white_check_mark:

File: dataprep-webapp/src/app/settings/actions/menu-actions-service.js [hunk 14]
```diff
@@ -25,8 +25,8 @@ export default class MenuActionsService {
 			break;
 		}
 		case '@@router/GO_DATASET': {
-			const { method, args, id } = action.payload;
-			this.$state[method](...args, { datasetid: id });
+			const { method, args, id, originalItem } = action.payload;
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/settings/actions/menu-actions-service.js]: I think we should adapt the use instead of adapting the logic here to introduce a particular case. Here we wait for an id in the payload.
* remove this `originalItem` variable
* in inventory-service
```
addHtmlLabelsAndSort(searchValue, data) {
		let inventoryItems = [];

		if (data.datasets && data.datasets.length) {
			_.each(data.datasets, function (item) {
				const itemToDisplay = {};

				itemToDisplay.id = item.id; <-- (add id here, it will be included in the payload)
				itemToDisplay.inventoryType = 'dataset';
				itemToDisplay.author = item.author;
				itemToDisplay.created = item.created;
				itemToDisplay.records = item.records;
				itemToDisplay.name = item.name;
				itemToDisplay.path = item.path;
				itemToDisplay.type = item.type;
				itemToDisplay.originalItem = item; <-- (here please name it model like the preps/datasets list)
				itemToDisplay.lastModificationDate = item.lastModificationDate;
				itemToDisplay.tooltipName = item.name;
				itemToDisplay.owner = item.owner;

				inventoryItems.push(itemToDisplay);
			});
		}
```

File: dataprep-webapp/src/app/components/search/inventory/inventory-search-controller.spec.js [hunk 15]
```diff
@@ -11,90 +11,71 @@
...
-        }));
-    });
+	let component;
+	let scope;
+
+	beforeEach(angular.mock.module('data-prep.inventory-search'));
+
+	beforeEach(inject(($rootScope, $componentController) => {
+		scope = $rootScope.$new();
+		component = $componentController('inventorySearch', { $scope: scope });
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/search/inventory/inventory-search-controller.spec.js]: :white_check_mark:

File: dataprep-webapp/src/app/components/search/inventory/inventory-search-controller.spec.js [hunk 16]
```diff
@@ -11,90 +11,71 @@
...
+			// given
+			spyOn(SearchService, 'searchAll').and.returnValue($q.when(results));
+
+			// when
+			component.search(searchInput);
+			component.currentInput = 'other';
+			scope.$digest();
+
+			// then
+			expect(component.results).not.toEqual(results);
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/search/inventory/inventory-search-controller.spec.js]: No array could not be checked with toBe in this case

File: dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js (hunk 8 above)
Comment (frassinier) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: :white_check_mark:

File: dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js (hunk 9 above)
Comment (frassinier) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: :white_check_mark:

File: dataprep-webapp/src/app/components/search/inventory/inventory-search-controller.spec.js [hunk 17]
```diff
@@ -11,90 +11,71 @@
...
+		}));
+
+		it('should set empty array as results when there are no result', inject(($q, SearchService) => {
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/search/inventory/inventory-search-controller.spec.js]: :white_check_mark:

File: dataprep-webapp/src/app/services/documentation/documentation-service.js [hunk 18]
```diff
@@ -20,56 +20,50 @@ class DocumentationService {
...
-	return _.chain(this._thcParser(response.data))
-                    .map((item) => {
-	this.textFormatService.highlight(item, 'name', keyword, 'highlighted');
-	this.textFormatService.highlight(item, 'description', keyword, 'highlighted');
-	return item;
-})
-                    .value();
-});
+			.then((response) => {
+				return _.chain(this._thcParser(response.data)).value();
```
Comment (frassinier) [inline dataprep-webapp/src/app/services/documentation/documentation-service.js]: :white_check_mark:

File: dataprep-webapp/src/app/services/search/search-module.js (hunk 11 above)
Comment (frassinier) [inline dataprep-webapp/src/app/services/search/search-module.js]: Sure :white_check_mark:

File: dataprep-webapp/src/app/settings/actions/menu-actions-service.js (hunk 14 above)
Comment (frassinier) [inline dataprep-webapp/src/app/settings/actions/menu-actions-service.js]: :white_check_mark:

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/72/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/72/)

**Comment by build-talend** [comment]
Pull request test result: [failure](http://ec2-54-209-208-131.compute-1.amazonaws.com:8080/job/tdp_pull-request_master/2/) [:x:](http://ec2-54-209-208-131.compute-1.amazonaws.com:8080/job/tdp_pull-request_master/2/)

**Comment by build-talend** [comment]
Pull request test result: [failure](http://ec2-54-209-208-131.compute-1.amazonaws.com:8080/job/tdp_pull-request_master/3/) [:x:](http://ec2-54-209-208-131.compute-1.amazonaws.com:8080/job/tdp_pull-request_master/3/)

**Comment by build-talend** [comment]
Pull request test result: [failure](http://ec2-54-209-208-131.compute-1.amazonaws.com:8080/job/tdp_pull-request_master/4/) [:x:](http://ec2-54-209-208-131.compute-1.amazonaws.com:8080/job/tdp_pull-request_master/4/)

**Comment by build-talend** [comment]
Pull request test result: [failure](http://ec2-54-209-208-131.compute-1.amazonaws.com:8080/job/tdp_pull-request_master/5/) [:x:](http://ec2-54-209-208-131.compute-1.amazonaws.com:8080/job/tdp_pull-request_master/5/)

File: dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js [hunk 19]
```diff
@@ -47,24 +79,119 @@ export default class AppHeaderBarCtrl {
...
+			...searchSettings,
+			icon: {
+				name: onToggleAction && onToggleAction.icon,
+				title: onToggleAction && onToggleAction.name,
+				bsStyle: 'link',
+			},
+			onToggle: this.searchOnToggle,
+			onBlur: this.searchOnBlur,
+			onChange: this.searchOnChange,
+			onSelect: this.searchOnToggle && this.searchOnSelect,
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: I don't understand, can you explain it ?

Comment (frassinier) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: I need to reset its render when I clicked on list item, otherwise I will keep search bar and its results displayed.

Comment (frassinier) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: It's interesting for folder selection

File: dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js [hunk 20]
```diff
@@ -47,24 +81,119 @@ export default class AppHeaderBarCtrl {
...
+					iconName: onSelectAction.icon,
+					iconTitle: onSelectAction.name,
+				});
+				onSelectDispatcherByType[type] = this.settingsActionsService.createDispatcher(onSelectAction);
+			}
+		});
+		this.searchOnSelect = (event, { sectionIndex, itemIndex }) => {
+			const selectedCategory = this.adaptedSearchResults[sectionIndex];
+			const selectedItem = selectedCategory && selectedCategory.suggestions[itemIndex];
+			const onSelectDispatcher = onSelectDispatcherByType[selectedItem.inventoryType];
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: Create the dispatcher outside of this callback that will be executed at each selection

File: dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js [hunk 21]
```diff
@@ -47,24 +81,119 @@ export default class AppHeaderBarCtrl {
...
+			this.searchOnToggle();
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/components/widgets-containers/app-header-bar/app-header-bar-controller.js]: `this.searchOnToggle && this.searchOnToggle()`

File: dataprep-webapp/src/app/services/search/search-service.js [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+			results = results.concat(response);
+		});
+
+		// if results (doc + inventory) are empty, we create an empty array
+		// the no-result message is based on the definition of results.
+		// It must be an empty array to show the message.
+		return $q
+			.all([inventoryPromise, documentationPromise])
+			.then(() => {
+				return results;
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/services/search/search-service.js]: Here you can simplify now : 
```
return $q
 			.all([inventoryPromise, documentationPromise])
 			.then([inventory, documentation] => inventory.concat(documentation);
```
and remove the 2 `then` and the local variable `results`

File: dataprep-webapp/src/app/services/search/inventory/search-inventory-service.js [hunk 23]
```diff
@@ -10,69 +10,72 @@
...
-     */
+	/**
+	 * @ngdoc method
+	 * @name addHtmlLabelsAndSort
+	 * @methodOf data-prep.services.search.inventory:SearchInventoryService
+	 * @param {String} searchValue string
+	 * @param {Object} data data to process
+	 * @description add html label to data based on searchValue and sort the results
+	 */
 	addHtmlLabelsAndSort(searchValue, data) {
```
Comment (jsomsanith) [inline dataprep-webapp/src/app/services/search/inventory/search-inventory-service.js]: searchValue is not used anymore (don't forget to remove it when it is called)
